### ISSUE TITLE
Timestamp not preserved when using -d option

### ISSUE THREAD

**Issue by tremby** [mjwoodcock/riscosarc#2]
I'm trying to uncompress a lot of old stuff at once. I want to uncompress but for the uncompressed version to have the same timestamp the compressed one had.

I was going to add `touch -r "$origfile" "${origfile%,d96}"` to my script to do this, before realizing that just chopping the `,d96` off the end won't do the trick, since I'm using `-x` to add another `,xxx` suffix, which I don't know at script run time. I'm having to do

    bash -O extglob -c 'riscosarc -x -F "$orig" && touch -r "$orig" "${orig%,d96}",!(d96)'

Using extglob always makes me nervous...

It'd be really nice to have an option which preserves the timestamp.

**Comment by mjwoodcock** [comment]
This would be achievable.

What timestamp do you currently get on the extracted files?  The current date, or the date of the file in the archive?

**Comment by tremby** [comment]
Oddly I've had mixed results. I've definitely seen extracted files come out with the current date (at least I think so. 2020 at least. I suppose this could have come from the date on the current directory or something?) but after I posted this ticket I also saw something extract to a 1993 date without me manually intervening. I can try to recreate...

**Comment by mjwoodcock** [comment]
No need to recreate, thanks.  I found a bug in the timestamp handling code.

I should be able to get this sorted for you tonight.

**Comment by tremby** [comment]
Ah. There's a bug here.

If I use the `-d` option to put the output in a particular directory, the file ends up with the current timestamp.

If I don't use the `-d` option, the file ends up with an older timestamp.

(My current test file is timestamped 1996, and when extracting without `-d` the resulting file ends up with a 1998 timestamp. I'm not sure why this would be. Surely if the file was last modified in 1998 it couldn't have been archived in 1996...? Maybe this is another bug, or maybe the timestamps on my files are wrong, either of the archives themselves or of the files inside?)

But here's the scarier part: in my tests just now I did something like

1. `riscosarc -x -F swashes,d96`
2. See that I now also have `swashes,d94`, the extracted file with a 1998 timestamp
3. `riscosarc -x -F -dtest swashes,d96`
4. See that I now have `test/swashes,d96` -- this one has a current timestamp
5. But now `swashes,d96` extracted in step 2 is gone!

This is a potentially-dangerous bug. If I write

    echo "important data" >some-file

and then

    riscosarc -x -dsomewhere-else some-file,d96

I would think that `some-file` in this directory is safe since any output will be written to `somewhere-else`. But in fact if `some-file` will be overwritten and then deleted!

**Comment by tremby** [comment]
Can you tell me what the bug you found is? Wondering if I need to re-extract some stuff to fix timestamps.

**Comment by mjwoodcock** [comment]
The bug is the one you found with the -d option: if you extract with the -d option, then timestamps will not be set on the extracted file.

The bug you found with deleting other files is a new one.  But I suspect I know what the issue is and can fix that easily.

**Comment by mjwoodcock** [comment]
OK, the deleting big is an easy fix.

I hope to get all three of your issues fixed tonight.

**Comment by tremby** [comment]
Thank you!

Any idea what might be happening with my 1996 vs 1998 timestamps?

**Comment by mjwoodcock** [comment]
It could well be a bug with the code that converts between RISC OS and Unix timestamps.  I can look at that.

Can you share the archive file for me to investigate?

**Comment by mjwoodcock** [comment]
If you can't share the archive, then "od -x [archive file] | head -2" will give me the timestamp of the archived file.

**Comment by tremby** [comment]
[swashes.zip](https://github.com/mjwoodcock/riscosarc/files/4615414/swashes.zip)

This won't let me upload the file unless I zip it, so it's in here. The archive has the timestamp 1996-11-10T17:05:29+0000 and when I extract it I get a file with timestamp 1998-02-10T19:19:33+0000. Would be interested to hear what you think.

**Comment by tremby** [comment]
```
0000000 1df8 0001 0303 0000 9447 fffd a699 1655
0000020 0101 000f de54 09c0 00a1 4140 7080 c888
```

**Comment by mjwoodcock** [comment]
Thanks.  I will take a look.

**Comment by mjwoodcock** [comment]
OK, there's a bug in my date conversion routine.

https://www.riscosopen.org/forum/forums/11/topics/10325 has an algorithm that gives a timestamp in 1996-10-01, which looks better.  So that should be a simple fix.

**Comment by tremby** [comment]
So that means the timestamps I applied to things with `touch -r` is right in this case, and so possibly others. That's a relief. Thanks!

I guess then that you'll make preserving the timestamp the default/only behaviour in all cases?

(I'll post a separate ticket for the file deletion so you can track its closure separately when you fix it!)

**Comment by mjwoodcock** [comment]
Yes, preserving the timestamp will be the default behaviour.  With a switch to use the archive file timestamp, and another switch to disable setting the timestamp.

**Comment by tremby** [comment]
Sounds very flexible!

**Comment by mjwoodcock** [comment]
I have pushed the changes to master.

By default, the timestamp of the file in the archive is used.

You can use the -D argument to prevent riscosarc from timestamping the file, and -T to use the timestamp of the archive file (I'm not sure you need this functionality any more, now the bug is fixed).

**Comment by tremby** [comment]
> You can use the -D argument to prevent riscosarc from timestamping
> the file,

Sounds great!

> and -T to use the timestamp of the archive file (I'm not sure you
> need this functionality any more, now the bug is fixed).

Nope, not any more, but I like the flexibility.

Thank you!

**Comment by tremby** [comment]
Seems to work perfectly. Thanks.

I thought at one point I had the same behaviour as `-d-` when not giving `-d` at all, but now I can't reproduce that. Not sure if I was imagining things.

**Comment by tremby** [comment]
I tested each of these. For a particular file (the one I uploaded above):

- The default leads to a file with timestamp 1996-10-01
- `-T` (make it match the archive) leads to 1996-11-10 which does indeed match the archive
- `-D` (don't set timestamp) leads to the current timestamp

This all seems as expected. Last editing the file on Oct 1 and then archiving it on Nov 10 of the same year seems realistic.

(I remember my dad always saying even when I was little that he was very careful when moving to new drives etc that timestamps were preserved, and I remember him talking about it again much more recently when he got his NAS. I've even seen textfile notes in some places on these disks where he's written things like "timestamp is wrong due to an accidental save. Really last edited on xxx date"! So I think the timestamps on the archive files can be trusted to be when the file was archived.)
